common/src/backends/auth.rs

```diff
@@ -3,17 +3,90 @@ use std::{future::Future, ops::Add, pin::Pin};
 use bytes::Bytes;
 use chrono::{Duration, Utc};
 use headers::{authorization::Bearer, Authorization, HeaderMapExt};
-use http::{Request, Response, StatusCode};
+use http::{Request, Response, StatusCode, Uri};
 use http_body::combinators::UnsyncBoxBody;
-use hyper::Body;
-use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header, Validation};
+use hyper::{body, Body, Client};
+use jsonwebtoken::{decode, encode, DecodingKey, EncodingKey, Header as JwtHeader, Validation};
 use serde::{Deserialize, Serialize};
 use tower::{Layer, Service};
-use tracing::error;
+use tracing::{error, trace};
+
+use super::headers::XShuttleAdminSecret;
 
 const EXP_MINUTES: i64 = 5;
 const ISS: &str = "shuttle";
 
+/// Layer to check the admin secret set by deployer is correct
+#[derive(Clone)]
+pub struct AdminSecretLayer {
+    secret: String,
+}
+
+impl AdminSecretLayer {
+    pub fn new(secret: String) -> Self {
+        Self { secret }
+    }
+}
+
+impl<S> Layer<S> for AdminSecretLayer {
+    type Service = AdminSecret<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AdminSecret {
+            inner,
+            secret: self.secret.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct AdminSecret<S> {
+    inner: S,
+    secret: String,
+}
+
+impl<S, ResponseError> Service<Request<Body>> for AdminSecret<S>
+where
+    S: Service<Request<Body>, Response = Response<UnsyncBoxBody<Bytes, ResponseError>>>
+        + Send
+        + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let error = match req.headers().typed_try_get::<XShuttleAdminSecret>() {
+            Ok(Some(secret)) if secret.0 == self.secret => None,
+            Ok(_) => Some(StatusCode::UNAUTHORIZED),
+            Err(_) => Some(StatusCode::BAD_REQUEST),
```

Comment:
Maybe we should use unauthorized if the header is missing too? I looked for a consensus on this myself when doing the auth layer, and it seems to be the convention. Auth0 says this: https://auth0.com/blog/forbidden-unauthorized-http-status-codes/#When-to-Use-401-Unauthorized

Reply:
Ah, nice. Yes, I agree. So the `Ok(None)` variant (second arm) means the header is missing. The `Err(_)` arm is when it failed to read the value in the header.